File: bindings/react-native/src/facets/graph.ts
import UldrenLoom from '../NativeUldrenLoom';
import { authArgs, keyArgs } from '../internal';
import type { LoomAuth, LoomKey } from '../internal';

/** Insert or replace node `id` (props as raw Loom Canonical CBOR `text -> bytes`, empty for none) in
 * graph `name` of `workspace` (created with the `graph` facet if absent). */
export function graphUpsertNode(
  loomPath: string,
  workspace: string,
  name: string,
  id: string,
  props: Uint8Array | number[],
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.graphUpsertNode(
    loomPath,
    workspace,
    name,
    id,
    Array.from(props),
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
}

/** Fetch node `id`'s props as raw Loom Canonical CBOR (`text -> bytes`), or null if absent. */
export async function graphGetNode(
  loomPath: string,
  workspace: string,
  name: string,
  id: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array | null> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  const bytes = await UldrenLoom.graphGetNode(
    loomPath,
    workspace,
    name,
    id,
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
  return bytes == null ? null : Uint8Array.from(bytes);
}

/** Remove node `id` from graph `name`; `cascade` also removes incident edges (else it conflicts while
 * any exist). */
export function graphRemoveNode(
  loomPath: string,
  workspace: string,
  name: string,
  id: string,
  cascade: boolean,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.graphRemoveNode(
    loomPath,
    workspace,
    name,
    id,
    cascade,
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
}

/** Insert or replace edge `id` from `src` to `dst` (both must exist) with `label` and props as raw
 * Loom Canonical CBOR `text -> bytes` (empty for none). */
export function graphUpsertEdge(
  loomPath: string,
  workspace: string,
  name: string,
  id: string,
  src: string,
  dst: string,
  label: string,
  props: Uint8Array | number[],
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.graphUpsertEdge(
    loomPath,
    workspace,
    name,
    id,
    src,
    dst,
    label,
    Array.from(props),
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
}

/** Fetch edge `id` as raw Loom Canonical CBOR `[src, dst, label, props]`, or null if absent. */
export async function graphGetEdge(
  loomPath: string,
  workspace: string,
  name: string,
  id: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array | null> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  const bytes = await UldrenLoom.graphGetEdge(
    loomPath,
    workspace,
    name,
    id,
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
  return bytes == null ? null : Uint8Array.from(bytes);
}

/** Remove edge `id` from graph `name`; resolves whether it was present. */
export function graphRemoveEdge(
  loomPath: string,
  workspace: string,
  name: string,
  id: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<boolean> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.graphRemoveEdge(
    loomPath,
    workspace,
    name,
    id,
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
}

/** Distinct adjacent node ids of `id` (sorted) as raw Loom Canonical CBOR (an array of text). */
export async function graphNeighborsCbor(
  loomPath: string,
  workspace: string,
  name: string,
  id: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.graphNeighbors(
      loomPath,
      workspace,
      name,
      id,
      passphrase,
      kek,
      authPrincipal,
      authPassphrase
    )
  );
}

/** Out-edges of `id` as raw Loom Canonical CBOR (an array of `[edge_id, edge]` in edge-id order). */
export async function graphOutEdgesCbor(
  loomPath: string,
  workspace: string,
  name: string,
  id: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.graphOutEdges(
      loomPath,
      workspace,
      name,
      id,
      passphrase,
      kek,
      authPrincipal,
      authPassphrase
    )
  );
}

/** In-edges of `id` as raw Loom Canonical CBOR (an array of `[edge_id, edge]` in edge-id order). */
export async function graphInEdgesCbor(
  loomPath: string,
  workspace: string,
  name: string,
  id: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.graphInEdges(
      loomPath,
      workspace,
      name,
      id,
      passphrase,
      kek,
      authPrincipal,
      authPassphrase
    )
  );
}

/** Node ids reachable from `start` as raw Loom Canonical CBOR (an array of text). `maxDepth < 0` is no
 * limit; `viaLabel` null/undefined follows every edge, else only edges with that label. */
export async function graphReachableCbor(
  loomPath: string,
  workspace: string,
  name: string,
  start: string,
  maxDepth: number,
  viaLabel: string | null | undefined,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.graphReachable(
      loomPath,
      workspace,
      name,
      start,
      maxDepth,
      viaLabel ?? null,
      passphrase,
      kek,
      authPrincipal,
      authPassphrase
    )
  );
}

/** A shortest path from `from` to `to` as raw Loom Canonical CBOR (an array of node-id text), or null
 * when none exists. `viaLabel` null/undefined follows every edge, else only edges with that label. */
export async function graphShortestPathCbor(
  loomPath: string,
  workspace: string,
  name: string,
  from: string,
  to: string,
  viaLabel: string | null | undefined,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array | null> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  const bytes = await UldrenLoom.graphShortestPath(
    loomPath,
    workspace,
    name,
    from,
    to,
    viaLabel ?? null,
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
  return bytes == null ? null : Uint8Array.from(bytes);
}

File: bindings/android/src/commonMain/kotlin/ai/uldren/loom/facets/graph.kt
package ai.uldren.loom

    /** Insert or replace node [id] in graph [name] of [workspace]; [props] is a CBOR `text -> bytes` map (empty = none). */
expect fun Loom.graphUpsertNode(
        path: String,
        workspace: String,
        name: String,
        id: String,
        props: ByteArray,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    )


    /** Fetch node [id]'s props as CBOR from graph [name] of [workspace], or null if absent. */
expect fun Loom.graphGetNode(
        path: String,
        workspace: String,
        name: String,
        id: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray?


    /** Remove node [id] from graph [name]; [cascade] true also removes incident edges. */
expect fun Loom.graphRemoveNode(
        path: String,
        workspace: String,
        name: String,
        id: String,
        cascade: Boolean,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    )


    /** Insert or replace edge [id] from [src] to [dst] with [label] and CBOR [props] in graph [name]. */
expect fun Loom.graphUpsertEdge(
        path: String,
        workspace: String,
        name: String,
        id: String,
        src: String,
        dst: String,
        label: String,
        props: ByteArray,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    )


    /** Fetch edge [id] as the CBOR array `[src, dst, label, props]`, or null if absent. */
expect fun Loom.graphGetEdge(
        path: String,
        workspace: String,
        name: String,
        id: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray?


    /** Remove edge [id] from graph [name]; returns whether it was present. */
expect fun Loom.graphRemoveEdge(
        path: String,
        workspace: String,
        name: String,
        id: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): Boolean


    /** Distinct adjacent node ids of [id] as a CBOR array of text. */
expect fun Loom.graphNeighborsCbor(
        path: String,
        workspace: String,
        name: String,
        id: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray


    /** Out-edges of [id] as a CBOR array of `[edge_id, edge]`. */
expect fun Loom.graphOutEdgesCbor(
        path: String,
        workspace: String,
        name: String,
        id: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray


    /** In-edges of [id] as a CBOR array of `[edge_id, edge]`. */
expect fun Loom.graphInEdgesCbor(
        path: String,
        workspace: String,
        name: String,
        id: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray


    /** Node ids reachable from [start] as a CBOR array of text; [maxDepth] < 0 = no limit, [viaLabel] null = any edge. */
expect fun Loom.graphReachableCbor(
        path: String,
        workspace: String,
        name: String,
        start: String,
        maxDepth: Long,
        viaLabel: String?,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray


    /** A shortest path from [from] to [to] as a CBOR array of node-id text, or null if none; [viaLabel] null = any edge. */
expect fun Loom.graphShortestPathCbor(
        path: String,
        workspace: String,
        name: String,
        from: String,
        to: String,
        viaLabel: String?,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray?
